File: macos/Gutenkern/Sources/GutenkernCore/KerningPlan.swift
import Foundation

public enum KerningGroup: String, CaseIterable, Hashable, Sendable {
    case capitals = "A"
    case smallCaps = "A.sc"
    case lowercase = "a"
    case punctuation = "-"
    case nonAlphabetic = "@"
    case liningFigures = "1p"
    case oldstyleFigures = "1o"
}

public enum KerningPlan {
    public static func rows(selected: Set<KerningGroup>) -> [[String]] {
        var result: [[String]] = []
        var currentBlock: KerningGroup?
        var currentItems: [String] = []
        for recipe in recipes {
            guard selected.contains(recipe.left), selected.contains(recipe.right) else {
                continue
            }
            if currentBlock != recipe.left {
                if !currentItems.isEmpty {
                    result.append(currentItems)
                }
                currentBlock = recipe.left
                currentItems = []
            }
            currentItems.append(recipe.line)
        }
        if !currentItems.isEmpty {
            result.append(currentItems)
        }
        return result
    }

    public static func text(selected: Set<KerningGroup>) -> String {
        rows(selected: selected).map { $0.joined(separator: " ") }.joined(separator: "\n")
    }

    private struct Recipe {
        let left: KerningGroup
        let right: KerningGroup
        let mode: PairMode

        var line: String {
            switch mode {
            case .simple:
                return "\(left.rawValue)/\(right.rawValue)"
            case .mirror:
                return "\(left.rawValue)/\(right.rawValue)/\(left.rawValue)"
            }
        }
    }

    private static let recipes: [Recipe] = [
        Recipe(left: .capitals, right: .capitals, mode: .mirror),
        Recipe(left: .capitals, right: .smallCaps, mode: .simple),
        Recipe(left: .capitals, right: .lowercase, mode: .simple),
        Recipe(left: .capitals, right: .punctuation, mode: .mirror),
        Recipe(left: .capitals, right: .nonAlphabetic, mode: .mirror),
        Recipe(left: .capitals, right: .liningFigures, mode: .mirror),
        Recipe(left: .capitals, right: .oldstyleFigures, mode: .mirror),

        Recipe(left: .smallCaps, right: .smallCaps, mode: .mirror),
        Recipe(left: .smallCaps, right: .punctuation, mode: .mirror),

        Recipe(left: .lowercase, right: .lowercase, mode: .mirror),
        Recipe(left: .lowercase, right: .punctuation, mode: .mirror),
        Recipe(left: .lowercase, right: .nonAlphabetic, mode: .mirror),
        Recipe(left: .lowercase, right: .liningFigures, mode: .mirror),
        Recipe(left: .lowercase, right: .oldstyleFigures, mode: .mirror),

        Recipe(left: .nonAlphabetic, right: .nonAlphabetic, mode: .mirror),
        Recipe(left: .nonAlphabetic, right: .liningFigures, mode: .mirror),
        Recipe(left: .nonAlphabetic, right: .oldstyleFigures, mode: .mirror),

        Recipe(left: .liningFigures, right: .liningFigures, mode: .mirror),
        Recipe(left: .liningFigures, right: .punctuation, mode: .mirror),

        Recipe(left: .oldstyleFigures, right: .oldstyleFigures, mode: .mirror),
        Recipe(left: .oldstyleFigures, right: .punctuation, mode: .mirror)
    ]
}

File: windows/Gutenkern.Core/KerningPlan.cs
namespace Gutenkern;

public enum KerningGroup
{
    Capitals,
    SmallCaps,
    Lowercase,
    Punctuation,
    NonAlphabetic,
    LiningFigures,
    OldstyleFigures
}

public static class KerningPlan
{
    public static string Code(KerningGroup group) => group switch
    {
        KerningGroup.Capitals => "A",
        KerningGroup.SmallCaps => "A.sc",
        KerningGroup.Lowercase => "a",
        KerningGroup.Punctuation => "-",
        KerningGroup.NonAlphabetic => "@",
        KerningGroup.LiningFigures => "1p",
        KerningGroup.OldstyleFigures => "1o",
        _ => throw new ArgumentOutOfRangeException(nameof(group), group, null)
    };

    public static bool TryParse(string code, out KerningGroup group)
    {
        switch (code)
        {
            case "A":
                group = KerningGroup.Capitals;
                return true;
            case "A.sc":
                group = KerningGroup.SmallCaps;
                return true;
            case "a":
                group = KerningGroup.Lowercase;
                return true;
            case "-":
                group = KerningGroup.Punctuation;
                return true;
            case "@":
                group = KerningGroup.NonAlphabetic;
                return true;
            case "1p":
                group = KerningGroup.LiningFigures;
                return true;
            case "1o":
                group = KerningGroup.OldstyleFigures;
                return true;
            default:
                group = default;
                return false;
        }
    }

    public static KerningGroup Parse(string code)
    {
        if (TryParse(code, out var group))
        {
            return group;
        }

        throw new ArgumentOutOfRangeException(nameof(code), code, null);
    }

    public static string Text(IEnumerable<KerningGroup> selected) =>
        string.Join("\n", Rows(selected).Select(row => string.Join(" ", row)));

    public static List<List<string>> Rows(IEnumerable<KerningGroup> selected)
    {
        var set = selected as HashSet<KerningGroup> ?? selected.ToHashSet();
        var result = new List<List<string>>();
        KerningGroup? currentBlock = null;
        var currentItems = new List<string>();
        foreach (var recipe in Recipes)
        {
            if (!set.Contains(recipe.Left) || !set.Contains(recipe.Right))
            {
                continue;
            }

            if (currentBlock != recipe.Left)
            {
                if (currentItems.Count > 0)
                {
                    result.Add(currentItems);
                    currentItems = [];
                }

                currentBlock = recipe.Left;
            }

            currentItems.Add(recipe.Line);
        }

        if (currentItems.Count > 0)
        {
            result.Add(currentItems);
        }

        return result;
    }

    public readonly record struct Recipe(KerningGroup Left, KerningGroup Right, PairMode Mode)
    {
        public string Line => Mode == PairMode.Simple
            ? $"{Code(Left)}/{Code(Right)}"
            : $"{Code(Left)}/{Code(Right)}/{Code(Left)}";
    }

    public static IReadOnlyList<Recipe> Recipes { get; } =
    [
        new(KerningGroup.Capitals, KerningGroup.Capitals, PairMode.Mirror),
        new(KerningGroup.Capitals, KerningGroup.SmallCaps, PairMode.Simple),
        new(KerningGroup.Capitals, KerningGroup.Lowercase, PairMode.Simple),
        new(KerningGroup.Capitals, KerningGroup.Punctuation, PairMode.Mirror),
        new(KerningGroup.Capitals, KerningGroup.NonAlphabetic, PairMode.Mirror),
        new(KerningGroup.Capitals, KerningGroup.LiningFigures, PairMode.Mirror),
        new(KerningGroup.Capitals, KerningGroup.OldstyleFigures, PairMode.Mirror),

        new(KerningGroup.SmallCaps, KerningGroup.SmallCaps, PairMode.Mirror),
        new(KerningGroup.SmallCaps, KerningGroup.Punctuation, PairMode.Mirror),

        new(KerningGroup.Lowercase, KerningGroup.Lowercase, PairMode.Mirror),
        new(KerningGroup.Lowercase, KerningGroup.Punctuation, PairMode.Mirror),
        new(KerningGroup.Lowercase, KerningGroup.NonAlphabetic, PairMode.Mirror),
        new(KerningGroup.Lowercase, KerningGroup.LiningFigures, PairMode.Mirror),
        new(KerningGroup.Lowercase, KerningGroup.OldstyleFigures, PairMode.Mirror),

        new(KerningGroup.NonAlphabetic, KerningGroup.NonAlphabetic, PairMode.Mirror),
        new(KerningGroup.NonAlphabetic, KerningGroup.LiningFigures, PairMode.Mirror),
        new(KerningGroup.NonAlphabetic, KerningGroup.OldstyleFigures, PairMode.Mirror),

        new(KerningGroup.LiningFigures, KerningGroup.LiningFigures, PairMode.Mirror),
        new(KerningGroup.LiningFigures, KerningGroup.Punctuation, PairMode.Mirror),

        new(KerningGroup.OldstyleFigures, KerningGroup.OldstyleFigures, PairMode.Mirror),
        new(KerningGroup.OldstyleFigures, KerningGroup.Punctuation, PairMode.Mirror)
    ];
}
